transport/http/handler_test.go

```diff
@@ -0,0 +1,190 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package http
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/yarpc/yarpc-go/encoding/raw"
+	"github.com/yarpc/yarpc-go/transport"
+	"github.com/yarpc/yarpc-go/transport/transporttest"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHandlerFailures(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	baseHeaders := make(http.Header)
+	baseHeaders.Set(CallerHeader, "somecaller")
+	baseHeaders.Set(EncodingHeader, "raw")
+	baseHeaders.Set(TTLMSHeader, "1000")
+	baseHeaders.Set(ProcedureHeader, "hello")
+	baseHeaders.Set(ServiceHeader, "fake")
+
+	headersWithBadTTL := headerCopyWithout(baseHeaders, TTLMSHeader)
+	headersWithBadTTL.Set(TTLMSHeader, "not a number")
+
+	tests := []struct {
+		req *http.Request
+		msg string
+	}{
+		{&http.Request{Method: "GET"}, "not found"},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, CallerHeader),
+			},
+			"caller name is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, ServiceHeader),
+			},
+			"service name is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, ProcedureHeader),
+			},
+			"procedure name is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, TTLMSHeader),
+			},
+			"ttlms is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headersWithBadTTL,
+			},
+			"ttlms must be an int",
+		},
+	}
+
+	for _, tt := range tests {
+		req := tt.req
+		if req.Body == nil {
+			req.Body = ioutil.NopCloser(bytes.NewReader([]byte{}))
+		}
+
+		h := handler{transporttest.NewMockHandler(mockCtrl)}
+		rw := httptest.NewRecorder()
+		h.ServeHTTP(rw, tt.req)
+
+		code := rw.Code
+		assert.True(t, code >= 400 && code < 500, "expected 400 level code")
```

Comment:
From response in original PR: 

Non-POST request is 404. Also, my intention here was to avoid imposing specific status codes on those failure cases. The only concrete requirement is that it's a 400-level failure.

Although, this will needs to be changed to follow https://github.com/yarpc/yarpc/blob/master/errors.md